src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -198,820 +542,1177 @@ private void checkDataLength(int ... lengths) {
+                    }
                 }
             }
-            // Prepare buffer for decryption
-            buffer.flip();
+            return 0;
         }
 
-        if (ct.remaining() > 0) {
-            ghashAllToS.doLastBlock(ct, ct.remaining());
-        }
-        // Prepare buffer for decryption if available
-        ct.reset();
+        /**
+         * Use any data from ibuffer and 'in' to first verify the auth tag. If
+         * the tag is valid, decrypt the data.
+         */
+        @Override
+        public int doFinal(byte[] in, int inOfs, int inLen, byte[] out,
+            int outOfs) throws IllegalBlockSizeException, AEADBadTagException,
+            ShortBufferException {
+            GHASH save = null;
+
+            int bufLen = getBufferedLength();
+            int len = inLen + bufLen;
+            if (len < tagLenBytes) {
+                throw new AEADBadTagException("Input too short - need tag");
+            }
 
-        byte[] block = getLengthBlock(sizeOfAAD, len);
-        ghashAllToS.update(block);
-        block = ghashAllToS.digest();
-        GCTR gctrForSToTag = new GCTR(embeddedCipher, this.preCounterBlock);
-        gctrForSToTag.doFinal(block, 0, tagLenBytes, block, 0);
+            if (len - tagLenBytes > out.length - outOfs) {
+                save = ghashAllToS.clone();
+            }
 
-        // check entire authentication tag for time-consistency
-        int mismatch = 0;
-        for (int i = 0; i < tagLenBytes; i++) {
-            mismatch |= tag.get() ^ block[i];
-        }
+            checkDataLength(len - tagLenBytes);
+            processAAD();
+
+            findTag(in, inOfs, inLen);
+            byte[] block = getLengthBlock(sizeOfAAD,
+                decryptBlocks(ghashAllToS, in, inOfs, inLen, null, 0));
+            ghashAllToS.update(block);
+            block = ghashAllToS.digest();
+            new GCTR(blockCipher, preCounterBlock).doFinal(block, 0,
+                tagLenBytes, block, 0);
+
+            // check entire authentication tag for time-consistency
+            int mismatch = 0;
+            for (int i = 0; i < tagLenBytes; i++) {
+                mismatch |= tag[i] ^ block[i];
+            }
 
-        if (mismatch != 0) {
-            throw new AEADBadTagException("Tag mismatch!");
-        }
+            if (mismatch != 0) {
+                throw new AEADBadTagException("Tag mismatch!");
+            }
 
-        // Decrypt the all the input data and put it into dst
-        doLastBlock(buffer, ct, dst);
-        restoreDst(dst);
-        src.position(src.limit());
-        // 'processed' from the gctr decryption operation, not ghash
-        return processed;
-    }
+            try {
+                ArrayUtil.nullAndBoundsCheck(out, outOfs, len - tagLenBytes);
+            } catch (ArrayIndexOutOfBoundsException aiobe) {
+                throw new ShortBufferException("Output buffer invalid");
+            }
```

Comment:
I had this down here because it's not needed until gctr ops are done and ghash doesn't use an output, but I can move it up.
I remember Sean C having to do save/restore work for GCM.. The tag can create the similar padding issues.  It felt safe to keep it.